File: tree/diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0
        res = 0
        left = get_length(root.left)
        right = get_length(root.right)
        curr = left + right
        return max(curr, self.diameterOfBinaryTree(root.left), self.diameterOfBinaryTree(root.right))

def get_length(node, res=0):
    if node is None:
        return 0
    return 1 + max(get_length(node.left), get_length(node.right))

File: tree/diameter-of-binary-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end

def get_depth(node)
    return 0 if node.nil?
    return 1 + [get_depth(node.left), get_depth(node.right)].max
end

# @param {TreeNode} root
# @return {Integer}
def diameter_of_binary_tree(root)
    return 0 if root.nil?
    left = get_depth(root.left)
    right = get_depth(root.right)
    res = left + right
    [res, diameter_of_binary_tree(root.left), diameter_of_binary_tree(root.right)].max
end